src/main.rs

```diff
@@ -150,9 +136,15 @@ async fn main() -> Result<(), Box<dyn Error>> {
     };
 
     let ctx = SnapchainValidatorContext::new(keypair.clone());
-    let block_proposer = BlockProposer::new(validator_address.clone(), shard.clone());
-    let shard_validator =
-        ShardValidator::new(validator_address.clone(), Some(block_proposer), None);
+    let block_proposer = Arc::new(Mutex::new(BlockProposer::new(
+        validator_address.clone(),
+        shard.clone(),
```

Comment:
We shouldn't share the proposer across threads. Whenever you end up requiring Arc/Mutex, we should think rethink how the dataflow works.


We should be using Message passing for concurrency instead of sharing memory: https://doc.rust-lang.org/book/ch16-02-message-passing.html

Take a look at how the block proposer is listening for confirmed shard chunks https://github.com/farcasterxyz/snapchain-v0/blob/2aa2393bc3b073fed82c2d5e2f73693ab0be59c2/src/consensus/consensus.rs#L292.


When @suurkivi is back we should talk about a higher level framework that will make this easier. I'm not a huge fan of ractor/actor model (it's very hard to follow the control flow), but it feels like we might end up rebuilding a lot of it if we try to do it from scratch.


In this case, I might do something like:
1. create a channel for the block proposer to communicate to the grpc server.
2. In register validator, send the highest height the validator has. And in consensus actor, when it receives the message, call proposer.is_ready_for_height (and start height should only be called when the proposer is ready)
3. If the proposer is missing blocks, send a message on the grpc server tx with a oneshot rx to receive the reply. For now you can just await on the rx, but we should push this to the background
4. When the gap is filled, it can return true and StartHeight